awx/main/tasks.py

```diff
@@ -372,16 +389,29 @@ def gather_analytics():
             if acquired is False:
                 logger.debug('Not gathering analytics, another task holds lock')
                 return
-            try:
-                tgz = analytics.gather()
-                if not tgz:
-                    return
-                logger.info('gathered analytics: {}'.format(tgz))
-                analytics.ship(tgz)
-                settings.AUTOMATION_ANALYTICS_LAST_GATHER = gather_time
-            finally:
-                if os.path.exists(tgz):
-                    os.remove(tgz)
+            subset = list(all_collectors().keys())
+            incremental_collectors = []
+            for collector in expensive_collectors():
+                if collector in subset:
+                    subset.remove(collector)
+                    incremental_collectors.append(collector)
+
+            # Cap gathering at 4 weeks of data if there has been no data gathering
+            since = last_time or (gather_time - timedelta(weeks=4))
+
+            if incremental_collectors:
+                start = since
+                until = None
+                while start < gather_time:
+                    until = start + timedelta(minutes=20)
+                    if (until > gather_time):
+                        until = gather_time
+                    if not _gather_and_ship(incremental_collectors, since=start, until=until):
+                        break
+                    start = until
+                    settings.AUTOMATION_ANALYTICS_LAST_GATHER = until
```

Comment:
@wenottingham hm, I think we were using DB `TowerAnalyticsState` table, because the collection can be made by a different host each time?

settings.AUTOMATION_ANALYTICS_LAST_GATHER will store it only on 1 host?

Reply:
hm I see, this is a specific settings implementation, not django settings.

So it seems like this will get saved to DB if the setting is part of `all_supported_settings` https://github.com/ansible/awx/blob/68c62d74f6dcbc1bfe49db5c209299f9f9905845/awx/conf/settings.py#L443 ?

Otherwise it gets stored only in memory.

Given we were using another DB table `TowerAnalyticsState`, I'd guess the AUTOMATION_ANALYTICS_LAST_GATHER change is stored only in memory :-)

Reply:
Hm, it looks like it is registered, so the change should be persisted to the DB https://github.com/ansible/awx/blob/68c62d74f6dcbc1bfe49db5c209299f9f9905845/awx/main/conf.py#L817

So the question is, why were we using another table to store this information? (it's either that it doesn't save properly to the DB or that the author didn't know about this :-) )

Reply:
I see, there was always
```
last_gather = Setting.objects.filter(key='AUTOMATION_ANALYTICS_LAST_GATHER').first()
```

But it was overridden later by 
```
state = TowerAnalyticsState.get_solo()	   
last_run = state.last_run
```

So that is quire strange :-o

Reply:
ok now I get it, so the original last run was the 
```
state = TowerAnalyticsState.get_solo()	   
last_run = state.last_run
```

The we added the setting as part of fixing the bug where all systems were sending the duplicate payload 
https://github.com/ansible/awx/pull/5721/files

Ok, so this should work fine (we can mark this as resolved)

Funny comment btw. :-D 

```
ADDITIONAL INFORMATION
This is probably abuse of the settings infrastructure to store the last collection time there.
```

Reply:
Yea, TowerAnalyticsState was the old thing that should have been removed when we first implemented the settings-based implementation of tracking the last run.